subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/javadoc/JavadocToolchainIntegrationTest.groovy

```diff
@@ -39,131 +47,219 @@ class JavadocToolchainIntegrationTest extends AbstractIntegrationSpec {
+        withInstallations(jdk1, jdk2).run(":javadoc")
         then:
         skipped(":javadoc")
     }
 
-    @IgnoreIf({ AvailableJavaHomes.getJdk(JavaVersion.VERSION_11) == null })
-    def "can manually set javadoc tool via  #type toolchain on javadoc task #type : #jdk"() {
-        buildFile << """
-            plugins {
-                id 'java'
-            }
+    def "fails on toolchain and executable mismatch"() {
+        def jdkCurrent = Jvm.current()
+        def jdkOther = AvailableJavaHomes.differentVersion
 
-            // need to do as separate task as -version stop javadoc generation
-            task javadocVersionOutput(type: Javadoc) {
-                source = tasks.javadoc.source
-                options.jFlags("-version")
-                javadocTool = javaToolchains.javadocToolFor {
-                    languageVersion = JavaLanguageVersion.of(${jdk.javaVersion.majorVersion})
-                }
-            }
+        configureProjectWithJavaPlugin()
 
-            javadoc {
-                javadocTool = javaToolchains.javadocToolFor {
-                    languageVersion = JavaLanguageVersion.of(${jdk.javaVersion.majorVersion})
-                }
-                dependsOn "javadocVersionOutput"
+        configureJavadocTool(jdkOther)
+        configureExecutable(jdkCurrent)
+
+        when:
+        withInstallations(jdkCurrent, jdkOther).runAndFail(":javadoc")
+
+        then:
+        failureHasCause("Toolchain from `executable` property does not match toolchain from `javadocTool` property")
+    }
+
+    def "fails on toolchain and executable mismatch (without java-base plugin)"() {
+        def jdkCurrent = Jvm.current()
+        def jdkOther = AvailableJavaHomes.differentVersion
+
+        configureProjectWithoutJavaBasePlugin()
+
+        configureJavadocTool(jdkOther)
+        configureExecutable(jdkCurrent)
+
+        when:
+        withInstallations(jdkCurrent, jdkOther).runAndFail(":javadoc")
+
+        then:
+        failureHasCause("Toolchain from `executable` property does not match toolchain from `javadocTool` property")
+    }
+
+    def "uses #what toolchain #when"() {
+        def jdkCurrent = Jvm.current()
+        def jdk1 = AvailableJavaHomes.differentVersion
+        def jdk2 = AvailableJavaHomes.getDifferentVersion(jdk1.javaVersion)
+
+        // When at least one toolchain is used for configuration, expect the first toolchain to be the target.
+        // Otherwise, expect the current toolchain as a fallback
+        def targetJdk = jdkCurrent
+        def useJdk = {
+            if (targetJdk === jdkCurrent) {
+                targetJdk = jdk1
+                return jdk1
+            } else {
+                return jdk2
             }
-        """
+        }
 
-        file('src/main/java/Lib.java') << testLib()
+        configureProjectWithJavaPlugin()
+
+        // Order of if's is important as it denotes toolchain priority
```

Comment:
Why does the order the `if`s are written matter here?  I don't understand.

Reply:
The reason the `if` order is important is because the `useJdk()` function calls actually modify some external state to make the tests work.

That makes me think that the tests should be slightly re-written to avoid this confusion. I'll address this shortly in the same PR.